GeneratorInterface/Core/interface/ConcurrentGeneratorFilter.h

```diff
@@ -299,24 +315,39 @@ namespace edm {
   }
 
   template <class HAD, class DEC>
-  void ConcurrentGeneratorFilter<HAD, DEC>::globalBeginLuminosityBlockProduce(LuminosityBlock& lumi,
-                                                                              EventSetup const& es) const {
-    // need one of the streams to finish
+  std::shared_ptr<gen::GenLumiCache<HAD, DEC>> ConcurrentGeneratorFilter<HAD, DEC>::globalBeginLuminosityBlock(
+      edm::LuminosityBlock const&, edm::EventSetup const&) const {
+    //need one of the streams to finish
     while (useInLumi_.load() == nullptr) {
     }
+
+    if (nInitializedInGlobalLumi_ < nGlobalBeginRuns_) {
+      while (!streamEndRunComplete_.load()) {
+      }
+      nInitializedInGlobalLumi_ = nGlobalBeginRuns_;
+    }
+
+    auto lumiCache = std::make_shared<gen::GenLumiCache<HAD, DEC>>();
+    lumiCache->useInLumi_ = useInLumi_.load();
```

Comment:
If `globalBeginLuminosityBlockProduce` uses the lumi cache's `useInLumi_` we could reset the member variable `useInLumi_` here which might help understanding the logic.

Reply:
It is critically important that streamBeginLuminosityBlock uses the value from the lumi cache. That is the main reason I created the lumi cache and only using it for that purpose emphasizes that requirement.

For globalBeginLuminosityBlockProduce, we could use either value (the module data member or the lumi cache member). To me it just seems easier to use the class data member because it avoids the extra code to get the value out of the lumi cache. I'll change this if you really want, but adding that seems less readable to me. globalBeginLuminosityBlock and globalBeginLuminosityBlockProduce just run one immediately after the other. If you really want me to change this, let me know and I will. I don't feel strongly about it.